File: 02_ProgrammingFundamentals/_11_RegularExpressions/_03_MoreExercises/_04_SantaSSecretHelper.java
package _02_ProgrammingFundamentals._11_RegularExpressions._03_MoreExercises;

import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class _04_SantaSSecretHelper {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        // Read input values
        int subtractNumber = Integer.parseInt(scan.nextLine());
        String input = scan.nextLine();

        // Read the message and decipher
        while (!input.equals("end")) {

            StringBuilder message = new StringBuilder();

            for (int i = 0; i < input.length(); i++) {
                message.append((char) (input.charAt(i) - subtractNumber));
            }

            String decodeRegex = "@(?<name>[A-Za-z]+)[^@\\-!:>]+(!)(?<grade>[GN]{1})(\\2)";
            Pattern decodePattern = Pattern.compile(decodeRegex);
            Matcher decodeMatcher = decodePattern.matcher(message);

            while(decodeMatcher.find()){
                if(decodeMatcher.group("grade").equals("G")){
                    System.out.println(decodeMatcher.group("name"));
                }
            }
            
            input = scan.nextLine();
        }
        
        scan.close();
    }
}

File: 24_C#Fundamentals/_09_RegularExpressions/_03_MoreExercises/_04_SantaSSecretHelper/_04_SantaSSecretHelper.cs
using System;
using System.Text.RegularExpressions;
using System.Text;
using System.Linq;

public class _04_SantaSSecretHelper
{
    public static void Main(string[] args)
    {
        // Read input values
        int subtractNumber = int.Parse(Console.ReadLine());
        string input = Console.ReadLine();

        // Read the message and decipher
        while (!input.Equals("end"))
        {
            StringBuilder message = new StringBuilder();

            for (int i = 0; i < input.Length; i++)
            {
                message.Append((char) (input[i] - subtractNumber));
            }

            string decodeRegex = @"@(?<name>[A-Za-z]+)[^@\-!:>]*!(?<grade>[GN])!";
            Regex decodePattern = new Regex(decodeRegex);
            MatchCollection decodeMatcher = decodePattern.Matches(message.ToString());

            foreach (Match match in decodeMatcher)
            {
                if (match.Groups["grade"].Value.Equals("G"))
                {
                    Console.WriteLine(match.Groups["name"].Value);
                }
            }

            input = Console.ReadLine();
        }
    }
}
